[#45799] Race condition with outdated OAuth access token

ba1ash opened the pull request:
https://community.openproject.org/work_packages/45799

Fixing race conditions when refreshing OAuth access tokens.

The OP front-end requests two api endpoints (file_links and storages) when a work package single view is loaded. The backend at times devides these two requests into two processes and each one requests a resources from NC. If the current OAuth token is outdated, both will try to refresh it using the same refresh token. One will finish earlier than the other. So the other will have by then an invalid refresh token. So, it is a race condition.

To mitigate this, this PR aims to enforce that the front-end emits the two requests strictly sequentially and not in parallel.

Also it fixes an inefficiency in the storages representer as it requested NC twice for the same authorization state check.

ba1ash committed:
[#45799] Race condition with outdated on OAuth access token

https://community.openproject.org/work_packages/45799

ba1ash committed:
Use Observable<void> instead of any type.

ba1ash committed:
Use an explicit typecast.

Kharonus commented on frontend/src/app/core/state/file-links/file-links.service.ts:
```diff
@@ -70,16 +72,14 @@ export class FileLinksResourceService extends ResourceCollectionService<IFileLin
-      );
+        }),
+        catchError(this.toastAndThrow.bind(this)),
+      ) as Observable<void>;
```
After a bit more thinking, I suggest a change:

- change return type to `Observable<IFileLink[]>`
- remove the error handling from here (remove `catchError`) and put the error handling (log and notify) into the caller
- add a `reduce` to the pipe, that aggregates the values of the observable stream.

Like it is now, this returned observable would emit value for every storage in the project of the work package. Here is the code suggestion:

```typescript
  updateCollectionsForWorkPackage(fileLinksSelfLink:string):Observable<IFileLink[]> {
    return this.http
      .get<IHALCollection<IFileLink>>(fileLinksSelfLink)
      .pipe(
        tap((collection) => insertCollectionIntoState(this.store, collection, fileLinksSelfLink)),
        switchMap((collection) => from(collection._embedded.elements)),
        groupBy(
          (fileLink) => fileLink._links.storage.href,
          (fileLink) => fileLink,
        ),
        mergeMap((group$) => {
          const seed = { storage: group$.key, fileLinks: [] as IFileLink[] };
          return group$.pipe(reduce((acc, fileLink) => {
            acc.fileLinks = [...acc.fileLinks, fileLink];
            return acc;
          }, seed));
        }),
        tap((fileLinkCollections) => {
          const storageId = idFromLink(fileLinkCollections.storage);
          const collectionKey = `${fileLinksSelfLink}?filters=[{"storage":{"operator":"=","values":["${storageId}"]}}]`;
          const collection = { _embedded: { elements: fileLinkCollections.fileLinks } } as IHALCollection<IFileLink>;
          insertCollectionIntoState(this.store, collection, collectionKey);
        }),
        reduce((acc, group) => acc.concat(group.fileLinks), [] as IFileLink[]),
      );
  }
```

ba1ash replied:
Applied, thanks!

Kharonus commented on frontend/src/app/features/work-packages/routing/wp-view-base/work-package-single-view.base.ts:
```diff
@@ -191,21 +191,23 @@ export class WorkPackageSingleViewBase extends UntilDestroyedMixin {
-    // Fetch file link collections for work package (only if storages module is enabled)
     if (this.workPackage.$links.fileLinks) {
-      this.fileLinkResourceService.updateCollectionsForWorkPackage(this.workPackage.$links.fileLinks.href as string);
+      this.fileLinkResourceService
+        .updateCollectionsForWorkPackage(this.workPackage.$links.fileLinks.href as string)
+        .subscribe(
+          () => {
+            this.projectsResourceService
+              .lookup((this.workPackage.project as unknown&{ id:string }).id)
+              .pipe(this.untilDestroyed())
+              .subscribe((project) => {
+                if (project._links.storages) {
+                  this.storages.updateCollection(project._links.self.href, project._links.storages);
+                }
+              });
+          },
+        );
```
To reduce nesting avoid a new observable in a subscribe, instead use something like this. And do not forget error handling here ;)

```typescript
    if (this.workPackage.$links.fileLinks) {
      this.fileLinkResourceService
        .updateCollectionsForWorkPackage(this.workPackage.$links.fileLinks.href as string)
        .pipe(
          this.untilDestroyed(),
          switchMap(() => this.projectsResourceService.lookup((this.workPackage.project as unknown&{ id:string }).id)),
        )
        .subscribe(
          (project) => {
            if (project._links.storages) {
              this.storages.updateCollection(project._links.self.href, project._links.storages);
            }
          },
          (error) => {
            this.toastService.addError(error);
            throw error;
          },
        );
    }
```

ba1ash replied:
Applied, thanks!

ba1ash committed:
Restructure the frontend code to be more readable.

- Respond with a list of FileLinks instead of void in FileLinksResourceService#updateCollectionsForWorkPackage
- Move error handling up to the caller of FileLinksResourceService#updateCollectionsForWorkPackage
- Avoid nested subscribe calls

Kharonus reviewed (approved):
LGTM

ba1ash committed:
Merge branch 'release/12.4' into bug/45799-race-condition-with-outdated-on-oauth-access-token
